src/raft/recv_install_snapshot.c

```diff
@@ -328,34 +279,482 @@ static const struct sm_conf leader_states[LS_NR] = {
+	assert(work != NULL);
+	*work = (struct raft_io_async_work) {
+		.work = create_ht_and_stmt,
+		.data = data,
+	};
+	state->io->async_work(work, create_ht_send_reply);
+}
+
+bool is_main_thread(void)
+{
+	// TODO: thread local storage.
 	return true;
 }
 
+void send_snapshot_cb(struct raft_io_send *req, int status) {
+	(void)req;
+	(void)status;
+	struct snapshot_leader_state *state = (struct snapshot_leader_state *) req->data;
+
+	sm_move(&state->sm, LS_FOLLOWER_NEEDS_SNAPSHOT);
+
+	raft_free(req);
+}
+
+void async_send_install_snapshot(struct snapshot_leader_state *state, const struct raft_message *msg) {
+	(void)msg;
+
+	struct raft_io_send *req;
+	req = raft_malloc(sizeof(*req));
+	assert(req != NULL);
+
+	req->data = (void *)state;
+
+	struct raft_message *reply;
+	reply = raft_malloc(sizeof(*reply));
+	assert(reply != NULL);
+
+	reply->type = RAFT_IO_INSTALL_SNAPSHOT;
+
+	state->io->async_send_message(req, reply, send_snapshot_cb);
+}
+
+void send_snapshot_done_cb(struct raft_io_send *req, int status) {
+	(void)req;
+	(void)status;
+	struct snapshot_leader_state *state = (struct snapshot_leader_state *) req->data;
+
+	state->ht_stmt = NULL;
+	sqlite3_close(state->ht);
+	state->ht = NULL;
+	sm_move(&state->sm, LS_SNAPSHOT_DONE_SENT);
+
+	raft_free(req);
+}
+
+void async_send_install_snapshot_done(struct snapshot_leader_state *state, const struct raft_message *msg) {
+	(void)msg;
+
+	struct raft_io_send *req;
+	req = raft_malloc(sizeof(*req));
+	assert(req != NULL);
+
+	req->data = (void *)state;
+
+	struct raft_message *reply;
+	reply = raft_malloc(sizeof(*reply));
+	assert(reply != NULL);
+
+	reply->type = RAFT_IO_INSTALL_SNAPSHOT;
+	reply->install_snapshot.result = DONE;
+
+	state->io->async_send_message(req, reply, send_snapshot_done_cb);
+}
+
+struct raft_message *get_signature_message(const struct snapshot_leader_state *state, const struct raft_message *msg) {
+	(void)msg;
+	(void)state;
+	struct raft_message *reply;
+	reply = raft_malloc(sizeof(*reply));
+	assert(reply != NULL);
+	// TODO: construct proper message with the next range of checksums required.
+	reply->type = RAFT_IO_SIGNATURE;
+
+	return reply;
+}
+
+void process_snapshot_installation_cp_or_mv(struct snapshot_leader_state *state, const struct raft_message *msg) {
+	if (msg->type == RAFT_IO_INSTALL_SNAPSHOT_CP_RESULT) {
+		state->last_page_acked = msg->install_snapshot_cp_result.last_known_page_no;
+	} else if (msg->type == RAFT_IO_INSTALL_SNAPSHOT_MV_RESULT) {
+		state->last_page_acked = msg->install_snapshot_mv_result.last_known_page_no;
+	}
+}
+
+void send_mv_or_cp_cb(struct raft_io_send *req, int status) {
+	(void)status;
+
+	struct snapshot_leader_state *state = (struct snapshot_leader_state *)req->data;
+	sm_move(&state->sm, LS_SNAPSHOT_CHUNCK_SENT);
+	raft_free(req);
+}
+
+void async_send_mv_or_cp(struct snapshot_leader_state *state) {
+	(void)state;
+	struct raft_message *reply;
+	struct raft_io_send *reply_req;
+
+	reply = raft_malloc(sizeof(*reply));
```

Comment:
Question:
What if we use macro declarations like these for allocations?
```
#define ALLOC_ARR(arr, nr)  calloc((nr), sizeof ((arr)[0]))
#define ALLOC_PTR(ptr)      ALLOC_ARR(ptr, 1)
```
cc @cole-miller

Reply:
```
ALLOC_PTR(reply);
```

Reply:
One issue with is is that we have a few different allocators that are used in different parts of the code, so until we clean that up (which I want to do) we'd need to duplicate these macros for each allocator.